step/step.go

```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
 func (step DockerBuildPushStep) Run() error {
 	var input Input
 	if err := step.inputParser.Parse(&input); err != nil {
-		return fmt.Errorf("failed to parse inputs: %w", err)
+		return fmt.Errorf("parse inputs: %w", err)
 	}
 	stepconf.Print(input)
 	step.logger.Println()
 
 	step.logger.EnableDebugLog(input.Verbose)
 
+	tagUsedInCacheKey := strings.Split(input.Tags, "\n")[0]
+
+	if input.UseBitriseCache {
+		if err := step.restoreCache(input, tagUsedInCacheKey); err != nil {
+			return fmt.Errorf("restore cache: %w", err)
+		}
+	}
+
+	if err := step.dockerBuild(input); err != nil {
+		return fmt.Errorf("build docker image: %w", err)
+	}
+
+	if input.UseBitriseCache {
+		if err := step.saveCache(input, tagUsedInCacheKey); err != nil {
+			return fmt.Errorf("save cache: %w", err)
+		}
+	}
+	return nil
+}
+
+func (step DockerBuildPushStep) restoreCache(input Input, imageName string) error {
+	step.logger.Infof("Restoring cache...")
+	saver := cache.NewRestorer(step.envRepo, step.logger, step.commandFactory)
+
+	var cacheKey = []string{
+		fmt.Sprintf(dockerCacheKeyTemplate, imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}-{{ .Branch }}", imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}", imageName),
```

Comment:
Listing these 3 keys tell me that maybe `imageName` should come before the other two variables and then a single cache key would be enough. Or did I miss something?

Reply:
What would the exact cache key look like in that case I am not sure I understand.

The reasoning for this here is to have a fallback hierarchy like so:

If we have built the same image already -> use that
otherwise check if we have built it already in the same branch
otherwise check if we have ever built it for this particular os-arch
otherwise this is the first time we are seeing it.